File: src/main/java/com/inttegro/demo/CheckoutService.java
package com.inttegro.demo;

import com.inttegro.ApiException;
import com.inttegro.Client;
import com.inttegro.RequestMeta;
import com.inttegro.customers.CustomerData;
import com.inttegro.money.Currency;
import com.inttegro.orders.CheckoutSettings;
import com.inttegro.orders.Order;
import com.inttegro.orders.OrderCreateParams;
import com.inttegro.orders.OrderLineItemParams;
import com.inttegro.prices.PriceParams;
import com.inttegro.products.ProductType;
import java.net.URI;

import org.springframework.stereotype.Service;

@Service
public class CheckoutService {
    /*
     * Inttegro integration map
     *
     * INTTEGRO:FLOW [hosted-checkout] This trusted service creates and finalizes
     * the Ledgerline Order and returns the hosted invoice URL.
     * INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY stays in the server
     * process. Never expose it through templates, static assets, client
     * configuration, logs, or exception responses.
     * INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers more UI
     * control but shifts more payment states, recovery, method-specific behavior,
     * testing, and compliance analysis to the merchant.
     * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
     * INTTEGRO:DOCS https://studio.inttegro.com/orders
     * INTTEGRO:DOCS https://studio.inttegro.com/keys
     * See the repository INTEGRATION_GUIDE.md and integration-decisions.json
     * for the shared rationale and machine-readable alternatives.
     */
    public record CheckoutResult(String orderId, String checkoutUrl) {}

    public static final class DemoException extends RuntimeException {
        private final String code;

        DemoException(String code, String message) {
            super(message);
            this.code = code;
        }

        public String code() { return code; }
    }

    static OrderCreateParams buildOrderRequest(CheckoutForm form, String origin) {
        return OrderCreateParams.builder()
                // INTTEGRO:DECISION [stable-idempotency-key] Reuse this key for
                // retries of one logical invoice payment. Production systems
                // should persist an invoice-derived key; a new key after a
                // timeout permits duplicates.
                // https://studio.inttegro.com/idempotency
                .requestMeta(RequestMeta.withIdempotencyKey("demo-" + form.attemptId()))
                // INTTEGRO:DECISION [inline-customer] customerData fits this
                // guest flow. Account-based apps should resolve customerId on
                // the server; Inttegro accepts exactly one representation.
                .customerData(CustomerData.builder().name(form.name().trim()).email(form.email().trim()).phoneNumber(form.phone().trim()).build())
                // INTTEGRO:DECISION [finalize-on-create] This invoice is settled,
                // so finalization freezes the Order and creates invoice formats.
                // Use draft -> update -> finalize when lines, tax, or approval
                // can still change.
                .finalizeOrder(true)
                .checkoutSettings(CheckoutSettings.builder().redirectUrl(origin + "/complete").cancelUrl(origin + "/cancel").build())
                .lineItem(OrderLineItemParams.product(product -> product
                        // INTTEGRO:DECISION [inline-product] Inline data keeps the
                        // demo independent of a catalog. Catalog merchants can
                        // use productId plus an explicit price or priceId; do not
                        // mix the catalog and inline shapes.
                        .type(ProductType.SERVICE)
                        .name("Invoice INV-2048")
                        .quantity(1)
                        // INTTEGRO:DECISION [minor-unit-money] Integer 5000 is
                        // GHS 50.00. Use a currency-aware decimal/money type to
                        // convert variable amounts.
                        .price(PriceParams.of(Currency.GHS, 5000))))
                .build();
    }

    CheckoutResult create(CheckoutForm form, String origin) {
        // INTTEGRO:SECURITY [server-api-key] Use the deployment secret manager
        // in production and rotate the key there; never serialize it to clients.
        String apiKey = System.getenv().getOrDefault("INTTEGRO_API_KEY", "").trim();
        if (apiKey.isEmpty()) {
            throw new DemoException("configuration_error", "Set INTTEGRO_API_KEY on the server.");
        }
        String publicOrigin = validatedOrigin(origin);
        try {
            // INTTEGRO:ALTERNATIVE [server-api-key] A production Spring app can
            // inject a singleton Client for transport reuse, fail-fast config,
            // and application-owned OpenTelemetry. Per-call creation is concise
            // here. https://studio.inttegro.com/sdk-observability
            Order order = new Client(apiKey).orders().create(buildOrderRequest(form, publicOrigin));
            // INTTEGRO:DECISION [returned-checkout-url] Use the response field.
            // Constructing a URL from order.id relies on undocumented routing.
            if (order.invoice == null || order.invoice.format == null || order.invoice.format.web == null || order.invoice.format.web.url == null) {
                throw new DemoException("api_error", "Inttegro did not return a hosted checkout URL.");
            }
            return new CheckoutResult(order.id, order.invoice.format.web.url);
        } catch (DemoException error) {
            throw error;
        } catch (ApiException error) {
            // INTTEGRO:SECURITY [safe-error-boundary] Keep raw responses,
            // credentials, and stack traces in protected diagnostics. Expose a
            // stable message and record only bounded metadata/request IDs.
            throw new DemoException("api_error", "Inttegro rejected the checkout request.");
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new DemoException("api_error", "Checkout was interrupted. Try again.");
        } catch (Exception error) {
            throw new DemoException("api_error", "Checkout is temporarily unavailable.");
        }
    }

    private static String validatedOrigin(String value) {
        // INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed
        // origin in production. A request-derived fallback is safe only behind
        // a precise trusted-proxy configuration. Return URLs must be HTTP(S),
        // and Inttegro appends order_id after checkout.
        try {
            URI uri = URI.create(value);
            if (!("http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) || uri.getHost() == null) throw new IllegalArgumentException();
            return uri.getScheme() + "://" + uri.getAuthority();
        } catch (IllegalArgumentException error) {
            throw new DemoException("configuration_error", "The demo public URL is invalid.");
        }
    }
}

File: rails/app/services/checkout_service.rb
require "inttegro"
require "uri"

# Inttegro integration map
#
# INTTEGRO:FLOW [hosted-checkout] This server-only service validates the Kora
# checkout, creates and finalizes an Order, and returns its hosted invoice URL.
# INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY belongs only in the trusted
# Rails process. Never expose it through credentials rendered to a browser,
# asset builds, client-side environment values, logs, or error responses.
# INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment provides more UI
# control but requires the merchant to own more payment states, recovery,
# method-specific behavior, testing, and compliance analysis.
# INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
# INTTEGRO:DOCS https://studio.inttegro.com/orders
# INTTEGRO:DOCS https://studio.inttegro.com/keys
# See ../../../INTEGRATION_GUIDE.md and ../../../integration-decisions.json for the
# shared rationale and machine-readable alternatives.

class CheckoutService
  Checkout = Data.define(:name, :email, :phone, :attempt_id)

  class DemoError < StandardError
    attr_reader :code

    def initialize(code, message)
      @code = code
      super(message)
    end
  end

  class << self
    def parse(values)
      checkout = Checkout.new(
        name: values.fetch("name", "").strip,
        email: values.fetch("email", "").strip,
        phone: values.fetch("phone", "").strip,
        attempt_id: values.fetch("attempt_id", "").strip
      )
      unless checkout.name.present? && checkout.phone.present? && checkout.email.match?(/\A[^\s@]+@[^\s@]+\.[^\s@]+\z/)
        raise DemoError.new("validation_error", "Enter a name, valid email, and phone number.")
      end
      unless checkout.attempt_id.match?(/\A[A-Za-z0-9_-]{8,100}\z/)
        raise DemoError.new("validation_error", "Start a fresh checkout and try again.")
      end
      checkout
    end

    def order_params(checkout, origin)
      {
        # INTTEGRO:DECISION [stable-idempotency-key] Reuse this key for a retry of
        # the same logical checkout. Production systems should persist a stable
        # cart/invoice-derived value; a new key after a timeout allows duplicates.
        # https://studio.inttegro.com/idempotency
        request_meta: { idempotency_key: "demo-#{checkout.attempt_id}" },
        # INTTEGRO:DECISION [inline-customer] customer_data fits this guest flow.
        # Account-based apps should resolve customer_id on the server; Inttegro
        # accepts exactly one of the two customer representations.
        customer_data: { name: checkout.name, email_address: checkout.email, phone_number: checkout.phone },
        # INTTEGRO:DECISION [finalize-on-create] This cart is already complete,
        # so one request freezes the order and creates invoice formats. Use a
        # draft, updates, then explicit finalization when tax, shipping,
        # inventory, approval, or line items can still change.
        finalize: true,
        checkout_settings: { redirect_url: "#{origin}/complete", cancel_url: "#{origin}/cancel" },
        line_items: [{
          type: "product",
          product: {
            # INTTEGRO:DECISION [inline-product] Inline product data makes this
            # demo self-contained. Catalog users can choose product_id with an
            # explicit price or price_id; never mix inline and catalog fields.
            type: Inttegro::ProductType::PHYSICAL,
            name: "Dawn Brew Set",
            quantity: 1,
            # INTTEGRO:DECISION [minor-unit-money] Integer 5000 means GHS 50.00.
            # Convert variable input with a decimal/money type, not Float.
            price: Inttegro::PriceParams.new(currency: Inttegro::Money::Currency::GHS, value: 5000)
          }
        }]
      }
    end

    def create(checkout, origin)
      # INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed
      # configured origin in production. A request-derived fallback is safe only
      # behind a precisely configured trusted proxy chain. Return URLs must be
      # absolute HTTP(S); Inttegro appends order_id on return.
      uri = URI.parse(origin)
      unless %w[http https].include?(uri.scheme) && uri.host
        raise DemoError.new("configuration_error", "The demo public URL is invalid.")
      end
      # INTTEGRO:SECURITY [server-api-key] Load this from the deployment secret
      # store and rotate it there; never serialize it to the browser.
      api_key = ENV.fetch("INTTEGRO_API_KEY", "").strip
      raise DemoError.new("configuration_error", "Set INTTEGRO_API_KEY on the server.") if api_key.empty?

      public_origin = "#{uri.scheme}://#{uri.host}"
      public_origin += ":#{uri.port}" unless [80, 443].include?(uri.port)
      # INTTEGRO:ALTERNATIVE [server-api-key] A production container can inject a
      # startup-configured client for transport reuse and application-owned
      # OpenTelemetry. Per-operation construction keeps this demo easy to trace.
      # https://studio.inttegro.com/sdk-observability
      order = Inttegro::Client.new(api_key: api_key).orders.create(**order_params(checkout, public_origin))
      # INTTEGRO:DECISION [returned-checkout-url] Use the response URL rather
      # than deriving one from order.id and undocumented routing conventions.
      checkout_url = order.invoice&.format_value&.web&.url
      raise DemoError.new("api_error", "Inttegro did not return a hosted checkout URL.") if checkout_url.blank?
      { order_id: order.id, checkout_url: checkout_url }
    rescue DemoError
      raise
    rescue Inttegro::APIError
      # INTTEGRO:SECURITY [safe-error-boundary] Raw SDK payloads, authorization
      # data, and stack traces stay in protected diagnostics. Public errors use
      # a stable vocabulary; log only bounded metadata and request IDs.
      raise DemoError.new("api_error", "Inttegro rejected the checkout request.")
    rescue StandardError
      raise DemoError.new("api_error", "Checkout is temporarily unavailable.")
    end
  end
end
